kf-protocol-derive/src/ast/container.rs

```diff
@@ -0,0 +1,78 @@
+use quote::ToTokens;
+use syn::{Attribute, Lit, Meta, NestedMeta};
+
+#[derive(Debug, Default)]
+pub struct ContainerAttributes {
+    pub varint: bool,
+    pub default: bool,
+    pub encode_discriminant: bool,
+    pub api_min_version: u16,
+    pub api_max_version: Option<u16>,
+    pub api_key: Option<u8>,
+    pub response: Option<String>,
+    pub repr_type_name: Option<String>,
+}
+
+impl ContainerAttributes {
+    pub fn from_ast(attributes: &[Attribute]) -> ContainerAttributes {
+        let mut cont_attr = ContainerAttributes::default();
+        // Find all supported container level attributes in one go
+        for attribute in attributes {
+            if attribute.path.is_ident("varint") {
+                cont_attr.varint = true;
+            } else if attribute.path.is_ident("fluvio_kf") {
+                if let Ok(Meta::List(list)) = attribute.parse_meta() {
+                    for kf_attr in list.nested {
+                        if let NestedMeta::Meta(Meta::NameValue(name_value)) = kf_attr {
+                            if name_value.path.is_ident("api_min_version") {
+                                if let Lit::Int(lit_int) = &name_value.lit {
+                                    cont_attr.api_min_version =
+                                        lit_int.base10_parse::<u16>().unwrap_or_else(|_| 0);
+                                }
+                            } else if name_value.path.is_ident("api_max_version") {
+                                if let Lit::Int(lit_int) = &name_value.lit {
+                                    cont_attr.api_max_version = lit_int.base10_parse::<u16>().ok();
+                                }
+                            } else if name_value.path.is_ident("api_key") {
+                                if let Lit::Int(lit_int) = &name_value.lit {
+                                    cont_attr.api_key = lit_int.base10_parse::<u8>().ok();
+                                }
+                            } else if name_value.path.is_ident("response") {
+                                if let Lit::Str(lit_str) = &name_value.lit {
+                                    cont_attr.response = Some(lit_str.value());
+                                }
+                            } else {
+                                log::warn!(
+                                    "#[fluvio_kf({})] does nothing on the container.",
+                                    name_value.to_token_stream().to_string()
```

Comment:
maybe then this should turn into compiler error?

Reply:
I wanted a warning, but I found no macro for that, like `compiler_error!`.

Reply:
However since this generally runs at compile time wouldn't a warning essentially be the same a simple `println!()`?

Reply:
ok.   leave as is